src/control/security/auth/auth_sys.go

```diff
@@ -146,67 +68,188 @@ func sysNameToPrincipalName(name string) string {
 	return name + "@"
 }
 
+func stripHostName(name string) string {
+	return strings.Split(name, ".")[0]
+}
+
 // GetMachineName returns the "short" hostname by stripping the domain from the FQDN.
 func GetMachineName() (string, error) {
 	name, err := os.Hostname()
 	if err != nil {
 		return "", err
 	}
 
-	return strings.Split(name, ".")[0], nil
+	return stripHostName(name), nil
+}
+
+type (
+	// CredentialRequest defines the request parameters for GetSignedCredential.
+	CredentialRequest struct {
+		DomainInfo      *security.DomainInfo
+		SigningKey      crypto.PrivateKey
+		getHostnameFn   func() (string, error)
+		getUserFn       func(string) (*user.User, error)
+		getGroupFn      func(string) (*user.Group, error)
+		getGroupIdsFn   func() ([]string, error)
+		getGroupNamesFn func() ([]string, error)
+	}
+)
+
+// NewCredentialRequest returns a properly initialized CredentialRequest.
+func NewCredentialRequest(info *security.DomainInfo, key crypto.PrivateKey) *CredentialRequest {
+	req := &CredentialRequest{
+		DomainInfo:    info,
+		SigningKey:    key,
+		getHostnameFn: GetMachineName,
+		getUserFn:     user.LookupId,
+		getGroupFn:    user.LookupGroupId,
+	}
+	req.getGroupIdsFn = func() ([]string, error) {
+		u, err := req.user()
+		if err != nil {
+			return nil, err
+		}
+		return u.GroupIds()
+	}
+	req.getGroupNamesFn = func() ([]string, error) {
+		groupIds, err := req.getGroupIdsFn()
+		if err != nil {
+			return nil, err
+		}
+
+		groupNames := make([]string, len(groupIds))
+		for i, gID := range groupIds {
+			g, err := req.getGroupFn(gID)
+			if err != nil {
+				return nil, err
+			}
+			groupNames[i] = g.Name
+		}
+
+		return groupNames, nil
+	}
```

Comment:
suggestion - To test the internals of these functions, I'd suggest breaking them out as normal named methods, and keeping the mockable interface as thin as possible.